### PR TITLE
[Site Name] Use site name value and skip domain name

### LINKED ISSUE #16249: [Site name] Submit the Site name value
## Description
The site name value entered by the user should be submitted to the API. For this a new parameter `blog_title` will be added in the existing API call `sites/new/` that is performed at the end of the flow.
([Android PR that removed the parameter](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1514/files#diff-921f1e3bd27968fb72c171fc9c0a3a29d8181e58df9d7e176df8e700d83d7d5eL219))

### LINKED ISSUE #16283: [Site name] Remove the domain screen from the flow
## Description
When the [Site Name screen is shown](https://github.com/wordpress-mobile/WordPress-Android/issues/16248) for the [experiment treatment variation](https://github.com/wordpress-mobile/WordPress-Android/issues/16246) the **site domain screen should be removed** from the site creation flow.

site-name-short

This modification adds two new use cases in the `sites/new` API call.
The domain of the site is generated with the following logic:
1. If the [siteName] is provided it is used as a domain (current behavior)
2. If the [siteName] is not provided the [siteTitle] is passed and the API generates the domain from it
3. If neither the [siteName] or the [siteTitle] is passed the [username] is used by the API to generate a domain

In the cases 2 and 3 two extra parameters are passed:
- `site_creation_flow` with value `with-design-picker`
- `find_available_url` with value `1`

Internal ref `pbArwn-49A-p2#comment-5390`

_Note: All development should be behind the [feature flag `SITE_NAME `](https://github.com/wordpress-mobile/WordPress-Android/issues/16245)

### PR DESCRIPTION
**Pull request by antonis**
Fixes #16249 and #16283

**Depends on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2344**

**WARNING**: [The `FluxCVersion`](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R16) needs to be updated with a reference to `trunk` after the `WordPress-FluxC-Android` PR has been merged.

## Description
This PR passes the site name value entered by the user to the API when the [Site Name screen is shown](https://github.com/wordpress-mobile/WordPress-Android/issues/16248) for the [experiment treatment variation](https://github.com/wordpress-mobile/WordPress-Android/issues/16246). It also skips the site domain screen from the site creation flow. In this case the domain of the site is generated from the site name or from the username when the site name is skipped.

To align with our discussions it also makes the following minor changes:
* [Capitalises first letter of the site name input](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/commits/3901ab465087a89c535c431dd2331c108a8adcba)
* [Removes the site name input hint](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/commits/85fed1f76b6857e98f821cadb627c292a82dde4a)

#### Prerequisites

<details>
<summary>Toggle the Site Intent Question feature</summary>

1. From My Site Go to Me → App Settings → Debug settings
2. Scroll to Features in development section and tap on SiteIntentQuestionFeatureConfig
3. Tap RESTART THE APP button
4. Since this feature is A/B tested the Site Intent Question will appear only to half of the users. Retry with a different account if it doesn't work

**Alternatively hardcode the feature on by returning `true` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/521e6b3a296339bec026f8ee2648eab326d31129/WordPress/src/main/java/org/wordpress/android/util/config/SiteIntentQuestionFeatureConfig.kt#L19)** 
</details>
<details>
<summary>Toggle the Site Name feature</summary>

1. From My Site Go to Me → App Settings → Debug settings
2. Scroll to Features in development section and tap on SiteNameFeatureConfig
3. Tap RESTART THE APP button
4. Since this feature is A/B tested go to **Abacus**, find the `wpandroid_site_name_v1` experiment and assign to your test a8c-user the treatment variation

**Alternatively hardcode the feature on by returning `true` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/521e6b3a296339bec026f8ee2648eab326d31129/WordPress/src/main/java/org/wordpress/android/util/config/SiteNameFeatureConfig.kt#L19)** 
</details>

### Current flow
1. Toggle the **Site Name** feature **OFF** 
2. Start the Site Creation flow
3. Verify that the Site Name screen is NOT shown
4. Select a design or skip
5. Verify that the Domain screen is shown
6. Search and select a domain and press continue
7. Verify that the site is created with the desired domain

https://user-images.githubusercontent.com/304044/162470264-d177468b-fdb3-421a-9471-d55830b17906.mp4

### Site name entered
1. Toggle the **Site Name** feature **ON** 
2. Toggle the **Site Intent Question** feature **ON** 
3. Start the Site Creation flow
4. Choose an intent for your site
5. Verify that the Site Name screen is shown
6. Entered your desired site name and press continue (e.g. My Site Name is Awesome)
7. Select a design or skip
9. Verify that the Domain screen is NOT shown
10. Verify that the site is created with a `*.wordpress.com` domain starting with name provided in 5 ( e.g. mysitenameisawesome) and possibly followed by a random number if the domain is used ( e.g. mysitenameisawesome13235)

https://user-images.githubusercontent.com/304044/162470383-bea2140f-0c5f-4faa-ad52-f568b151383e.mp4

### Site name skipped
1. Toggle the **Site Name** feature **ON** 
2. Toggle the **Site Intent Question** feature **ON** 
3. Start the Site Creation flow
4. Choose an intent for your site
5. Verify that the Site Name screen is shown
6. Press the Skip button
7. Select a design or skip
8. Verify that the Domain screen is NOT shown
9. Verify that the site is created with a `*.wordpress.com` domain starting with your username ( e.g. myusername) and possibly followed by a random number if the domain is used ( e.g. myusername13235)

https://user-images.githubusercontent.com/304044/162470487-b64853b3-764f-47bc-9004-d7f672310ece.mp4

## Regression Notes
1. Potential unintended areas of impact
Site creation flow

12. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests and the following unit test: [SitePreviewVieModelTests](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/commits/b4a4712cc00c87ff0f2d834d8cf5e4b18d678561), [Updates SiteCreationServiceManagerTests](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/commits/cdcaa90b99024c411df9ed14498f4671d71d3691) and [CreateSiteUseCaseTests](https://github.com/wordpress-mobile/WordPress-Android/pull/16278/commits/a3c3956aebebcb6b2759d52381484fe2989a9b1c)

13. What automated tests I added (or what prevented me from doing so)
Enhanced the unit tests mentioned above(2)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt]: What is the scenario where both of these would be null? Is this something we can still handle as a payload (as `""`), or an error state of some kind? If the latter, should we log it?

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt]: I'm curious when we'd have an email, but not have the username :thinking: . Is this mainly relevant for site creation during signup? If so, does creating the domain based on email have any privacy implications - enough to warrant a dialog / disclosure of some sort? Otoh, I understand we are trying to simplify the experience, so introducing additional messaging is a trade-off to consider :sweat_smile:

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SitePreviewViewModel.kt
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SitePreviewViewModel.kt:138]: This line relates with a [review comment I left on the companion PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2344#discussion_r846945483).

**Review by mkevins — COMMENTED** [review]
This is great work Antonis! I tested this, and it is working as described. I've left a few comments / questions / suggestions.

Also, thank you for adding tests, and for addressing the additional items updated from the discussion :+1:

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt]: I've investigated this further and I don't see any cases where this would be null. The logic is updated with 31301744d733d18fee46f4a2a31fd435140f34e4

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt]: Great catch @mkevins 👍 Thank you for bringing this up 🙇 
I got into using the email [while checking this other usage of the getuserName()](https://github.com/wordpress-mobile/WordPress-Android/blob/082b8f039aef4abf647e9b927bd3d3834137f25e/WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupEpilogueFragment.java#L632) function. This is a completely different usecase though since it is the epilogue of the signup where the username is generated from the user provided email. My case is not the same since it is not used in a user facing form but the public url which is not a good idea since it creates a privacy issue 😬 
Investigating this further I came to the conclusion that the user will always have a username at this point and there is no need for the added logic.
I updated this with 31301744d733d18fee46f4a2a31fd435140f34e4

**Comment by antonis** [comment]
Thank you for reviewing this @mkevins and providing precious feedback 🙇 
I've made some changes with 31301744d733d18fee46f4a2a31fd435140f34e4 tackling the reported issues.

**Review by ovitrif — COMMENTED** [review]
Nice work @antonis!

I reviewed the changes and everything LGTM while waiting for @mkevins 's feedback on your answers on his comments.

Testing results were all good for every scenario:

- 🟢 Current flow
- 🟢 Site name entered + different explorations trying to break this
- 🟢 Site name skipped + different explorations trying to break this

✅ to :shipit: from my side 🎉 

P.S. Approved since I see no blocker, but we can wait for more feedback from Matthew before merging 🙇 , since we'll have to update the reference to WordPress-FluxC-Android version once https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2344 is merged.

**Review by ovitrif — APPROVED** [review]

**Review by mkevins — APPROVED** [review]
Thanks for updating this Antonis! Tested again, and it is working as described :tada:

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/usecases/CreateSiteUseCase.kt]: Thanks for updating to address this :heart:

**Pull request merged**
